### ISSUE TITLE
On the Effect of Pretraining Corpora on In-context Learning by a Large-scale Language Model

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#28]
(https://engineering.clova.ai/posts/2022/05/hyperclova-corpus?fbclid=IwAR1_defvQG2s4jP3f6CyvvOvw5FGwF_0WcPyHSz8ITU2sEe7afeTE-qJqFs)
# 개요

- 대규모 언어 모델의 in-context zero, few-shot learning 능력이 언제 생기는지에 대한 연구가 부족함.
    - 예컨대, 학습 코퍼스가 달라지면 in-context learning 성능이 어떻게 달라지는지 등
- 본 논문에서는 Korean-GPT-3(HyperCLOVA)를 이용하여 사전학습 코퍼스의 소스와 사이즈에 따른 in-context learning 능력을 조사함.
    - 아래와 같은 관찰 결과를 리포트함.
        1. in-context learning 성능은 코퍼스 도메인 소스에 크게 의존하고, 코퍼스의 사이즈는 반드시 in-context learning의 emergence를 결정짓지 않는다.
        2. in-context learning 성능은 언어 모델이 다양한 코퍼스의 조합으로 학습되었을 때 생길 수 있다.
        3. 다운스트림 태스크와 관련있는 코퍼스로 사전학습하는 것이 반드시 뛰어난 다운스트림 in-context learning 성능을 보장해주지는 않는다.
        4. 언어 모델링과 in-context learning은 큰 상관성이 있지 않다.

# 태스크 정의

## 모델

- 파라미터 사이즈와 사전학습 코퍼스가 각각 다른 HyperCLOVA의 variants를 사용함.
    - 모델은 주로 1.3B 모델을 이용 (6.9B의 성능도 같이 리포트)
    - 모든 모델의 최대 시퀀스 길이는 2048
    - 모든 모델은 morpheme-aware byte-level BPE tokenizer 이용

## 다른 코퍼스로 사전학습

- 7개의 subcorpora에 대한 영향을 분석
    - Blog, Cafe, News, Comments, KiN, Modu, Ency
        
<img width="546" alt="1" src="https://user-images.githubusercontent.com/49383323/188169373-a085e012-d92e-45d7-86d1-beb35cf03b58.png">
        
- 150B 토큰보다 적은 corpora에 대해서, 99%는 사전학습에 이용하고 남은 1% 중 10,000개 예시를 랜덤 샘플링하여 validation ppl을 측정하는데에 이용.
- 150B 토큰보다 많은 corpora에 대해서는, 랜덤 샘플링을 통해 토큰 수를 150B로 조정하고, 남은 파트 중 10,000개를 validation 용으로 이용함
    - 각 코퍼스의 최대 학습셋 사이즈를 150B로 맞춤.
- 총 7만개의 검증 셋은 early stopping을 위해 사용.

## Downstream Tasks

- in-context learning 성능을 평가하기 위해 4가지 다운스트림 태스크를 이용
    - NSMC: 영화 감성 리뷰
    - KorQuAD: MRC
    - AI Hub translation: 한영 번역
    - YNAT: 토픽 7진 분류
- 4가지 태스크중 3가지 태스크가 사전학습시에 사용하는 코퍼스와 유사한 성격을 지님
    - KorQuAD의 passages는 Korean Wikipedia 기반이기 때문에 **Ency**와 연관이 있고
    - YNAT은 뉴스의 헤드라인의 토픽 분류 문제이기 때문에 **News**와 연관이 있고
    - AI Hub translation 태스크는 뉴스 기사에서 유래하였기 때문에 **News**와 연관이 있고, KiN 코퍼스의 2.5%는 한영 번역 관련 질문이기 때문에 **KiN**과도 연관이 있음
        - (다운) 생각해보니 KiN이 정말 양질의 코퍼스일 듯

## Experimental Details

- 모델은 1024 배치사이즈로 72K 스텝 학습함.
    - 이 세팅으로 학습하면 1024*72000*2048=150B 정도의 토큰을 학습하게 되기 때문에, 코퍼스의 최대 크기를 150B로 설정함.
- 대부분의 실험에서 validation perplexity가 단조감소하기 때문에 72K 스텝의 체크포인트를 이용함.
    - Ency 모델은 유일하게 12K에서 검증 로스가 제일 낮았음.
        - 데이터가 적어서 그런듯
- AdamW with lr 2.0e-4, cosine learning rate scheduling
- 모델은 A100 1024개로 구성된 Nvidia Superpod으로 학습함.
    - 1.3B 모델을 72K 스텝 학습하는데에 18시간 소요되었다고
- NSMC, YNAT 등의 분류 문제는 rank classification approach로 치환하여 품
    - pre-defined outputs (”positive”, “negative”)를 비교하여 더 확률이 높은 쪽을 선택
- KorQuAD, AIHub는 그리디 디코딩으로 아웃풋 토큰을 생성
- few-shot 실험에서 shots의 개수
    - 70: NSMC
    - 4: KorQuAD
    - 4: AIHub
    - 70: YNAT
- 다운스트림 태스크는 각각 12, 1, 3, 6번씩 랜덤 시드를 바꾸며 수행
    - 리포트 결과는 해당 실험 결과들의 평균

## Measuring Validation Perplexity

- 검증셋은 7개의 코퍼스 소스로부터 추출한 총 7만건의 예제로 구성됨.
- PPL 계산 시 모든 실험은 동일 vocabulary, validation set을 이용
- 두개 이상의 혼합된 corpora로 학습한 모델을 Italic 폰트로 기재함.
- 멀티 도메인 모델은 단일 도메인 모델보다 일반적으로 PPL이 더 낮을 것.

# Experimental Results

- 4개의 메인 질문
    1. 사전학습 corpora의 소스와 사이즈가 in-context learning 능력이 부상하는데 어떤 영향을 줄 것인가?
    2. 다양한 corpora를 혼합했을 때의 영향은 무엇인가?
    3. 다운스트림 태스크의 성능에 코퍼스의 도메인 연관성이 얼마나 큰 영향을 미칠 것인가?
    4. validation PPL과 in-context learning의 성능이 얼마나 큰 상관관계를 가질 것인가?

## Main Results

<img width="1108" alt="2" src="https://user-images.githubusercontent.com/49383323/188169552-bda21868-d366-4a1f-b7dc-05752b3f2a57.png">
<img width="854" alt="3" src="https://user-images.githubusercontent.com/49383323/188169562-360dc923-9e71-4d11-9084-7d23eb24edb0.png">
<img width="867" alt="4" src="https://user-images.githubusercontent.com/49383323/188169565-e801cde2-1fad-4d7e-9fc0-52c1293bc6f4.png">
<img width="835" alt="5" src="https://user-images.githubusercontent.com/49383323/188169568-62df8de0-a019-4cac-b1dc-54776274173f.png">
<img width="463" alt="6" src="https://user-images.githubusercontent.com/49383323/188169571-61341870-df7a-47a5-99dd-6b1185acbf88.png">

- 특정 코퍼스로만 학습했을 때 in-context 능력이 생기지 않는 경우가 존재함.

## Effect of Corpus Source

- Blog 코퍼스로 학습한 모델은 ALL 모델과 비슷한 수준의 in-context learning 성능을 보여줌.
- 반면에, Cafe, News 코퍼스로 학습한 모델은 in-context learning 성능이 박살
    - 심지어 Modu 코퍼스는 Cafe, News 보다 코퍼스 사이즈가 훨씬 작음에도 불구하고 in-context 성능이 어느정도 나옴.
    - **코퍼스 사이즈가 in-context learning 성능의 유일한 팩터는 아니다.**
- Cafe+News 코퍼스로 학습한 모델 역시 in-context learning 성능이 박살

## Effect of Corpus Size

- Table 6를 보면 코퍼스 사이즈를 150B→56B로 낮춰도 성능이 크게 감소하지 않음.
- 그러나 6B로 넘어가면 성능이 많이 떨어짐.
    - 그럼에도 불구하고 Cafe+News의 150B보단 나은 성능.
- Table 2를 보면 Blog 27B<<Blog 54B $\approx$ Blog 150B

## Effect of Combining Corpora

- Table 4를 보면 KiN과 Ency를 각각 학습시킨 모델은 in-context learning 능력이 없지만, KiN+Ency로 학습한 모델은 in-context learning 능력이 생김.
    - Cafe+KiN도 마찬가지.
- **in-context learning 능력은 multi-task learning에 의해 발현한다.**
    - next word를 예측하는 LM 태스크 특성상 다양한 도메인의 다음 단어를 예측하는 것이 일반화 성능 향상에 큰 도움이 되지 않을까?
- combining한다고 반드시 in-context learning 능력이 생기는건 또 아님ㅎ
    - Cafe+News로는 안생기더라

## Effect of Domain Relevance

- Table 2를 보면 KiN과 Ency가 MRC 태스크에서 좋은 성능을 보이지는 않음.
    - Ency는 KorQuAD가 한국어 Wikipedia 기반으로 만들어진 데이터셋
    - KiN는 question answering pairs
- News가 YNAT 태스크에서 좋은 성능을 보이지는 않음.
    - YNAT은 뉴스 헤드라인 쿼리로 구성된 데이터셋
- Table 4를 보면 YNAT 태스크에 대해서 News+KiN+Ency<KiN+Ency
- **다운 스트림 태스크와 사전학습 코퍼스의 연관성이 반드시 좋은 in-context learning 성능을 보장하진 않는다.**

<img width="474" alt="7" src="https://user-images.githubusercontent.com/49383323/188169620-ca8bc7e1-730f-4a25-a208-f3eef50f5c5f.png">

- Modu 코퍼스는 Ai Hub와 vocab overlap 비율이 높지만, Blog와 KiN보다도 번역 태스크 성능이 떨어짐.
- 반면에, zero-shot 성능은 도메인 관련성과 positive한 상관관계를 띄는 것으로 보임.

## Perplexity and Downstream Task

<img width="482" alt="8" src="https://user-images.githubusercontent.com/49383323/188169646-e479bf08-adf6-40e3-a3de-2d29096ec808.png">

- valid PPL과 downstream task 성능 간에 특별한 경향성을 찾기 어려움.
